### PR TITLE
Use groups the Slack bot/workflow has been whitelisted for when talki…

### PR DESCRIPTION
**Pull request by lasryaric**
Use groups the Slack bot/workflow has been whitelisted for when talking to the Dust API.

Task is [here](https://github.com/dust-tt/dust/issues/6664).

### CODE REVIEW

File: connectors/src/connectors/slack/bot.ts [hunk 1]
```diff
@@ -246,11 +246,20 @@ async function botAnswerMessage(
     throw new Error("DUST_FRONT_API environment variable is not defined");
   }
 
+  let requestedGroups: string[] | undefined = undefined;
+
+  if (slackUserInfo.is_bot) {
+    const botName = slackUserInfo.real_name || slackUserInfo.display_name;
+    if (typeof botName === "string") {
```
Comment (flvndvd) [inline connectors/src/connectors/slack/bot.ts:253]: This feels weird 🤔. What's the type of `botName`?

File: connectors/src/connectors/slack/bot.ts [hunk 2]
```diff
@@ -246,11 +246,20 @@ async function botAnswerMessage(
     throw new Error("DUST_FRONT_API environment variable is not defined");
   }
 
+  let requestedGroups: string[] | undefined = undefined;
+
+  if (slackUserInfo.is_bot) {
+    const botName = slackUserInfo.real_name || slackUserInfo.display_name;
```
Comment (flvndvd) [inline connectors/src/connectors/slack/bot.ts:252]:
```suggestion
    const botName = slackUserInfo.real_name ?? slackUserInfo.display_name;
```

File: types/src/front/lib/dust_api.ts [hunk 3]
```diff
@@ -137,6 +136,9 @@ export type DustAppRunFunctionCallArgumentsTokensEvent = {
 export type DustAPICredentials = {
   apiKey: string;
   workspaceId: string;
+  // Dust system API keys can request to be authenticated as a list of specific group IDs.
+  // This is for internal use only.
+  groups?: string[];
```
Comment (flvndvd) [inline types/src/front/lib/dust_api.ts:141]:
```suggestion
  groupIds?: string[];
```

File: types/src/front/lib/dust_api.ts [hunk 4]
```diff
@@ -379,8 +380,8 @@ export class DustAPI {
       "Content-Type": "application/json",
       Authorization: `Bearer ${this._credentials.apiKey}`,
     };
-    if (groups.length > 0) {
-      headers[DustGroupIdsHeader] = groups.map((g) => g.sId).join(",");
+    if (this._credentials.groups) {
```
Comment (flvndvd) [inline types/src/front/lib/dust_api.ts:383]: It feels weird to have this logic hidden and implicit. I think it's the type of logic that even outside of this module you should be aware of, WDYT?

File: connectors/src/connectors/slack/bot.ts (hunk 2 above)
Comment (lasryaric) [reply connectors/src/connectors/slack/bot.ts:252]: Honestly I never understand why ?? is better than || in this case. Do we have a convention in the code base? I mean here it really means one or the other.

File: types/src/front/lib/dust_api.ts (hunk 4 above)
Comment (lasryaric) [reply types/src/front/lib/dust_api.ts:383]: Agree that's one way to see it, but for me this is extremely close to being a credential, so I think it's the right place.

File: connectors/src/connectors/slack/bot.ts (hunk 1 above)
Comment (lasryaric) [reply connectors/src/connectors/slack/bot.ts:253]: Indeed, that weird, I did not want to challenge the type `SlackUserInfo`, and we do that when we check in the database if the bot has been whitelisted, but I changed the type `SlackUserInfo` to make `real_name` non nullable, and we'll use that field to check on the bot name from now on. Pushing a commit soon.

File: connectors/src/connectors/slack/lib/slack_client.ts [hunk 5]
```diff
@@ -142,6 +142,10 @@ export async function getSlackUserInfo(
     throw res.error;
   }
 
+  if (!res.user?.profile?.real_name) {
```
Comment (flvndvd) [inline connectors/src/connectors/slack/lib/slack_client.ts:145]: Assuming we know that it's always defined? 

Side note but if Slack's module types are too loose, io-ts could be a great solution here.

File: connectors/src/connectors/slack/lib/workspace_limits.ts [hunk 6]
```diff
@@ -206,10 +206,7 @@ export async function isBotAllowed(
   connector: ConnectorResource,
   slackUserInfo: SlackUserInfo
 ): Promise<Result<undefined, Error>> {
-  const displayName = slackUserInfo.display_name ?? "";
-  const realName = slackUserInfo.real_name ?? "";
-
-  const names = [displayName, realName].filter((name) => name.length > 0);
+  const realName = slackUserInfo.real_name;
```
Comment (flvndvd) [inline connectors/src/connectors/slack/lib/workspace_limits.ts:209]: Just want to make sure that this isn't a breaking change and that we test before releasing.

File: front/lib/actions/server.ts [hunk 7]
```diff
@@ -86,18 +86,14 @@ export async function runActionStreamed(
   const now = new Date();
 
   const prodCredentials = await prodAPICredentialsForOwner(owner);
+  const requestedGroupIds = auth.groups().map((g) => g.sId);
   const api = new DustAPI(
     apiConfig.getDustAPIConfig(),
-    prodCredentials,
+    { ...prodCredentials, groups: requestedGroupIds },
```
Comment (flvndvd) [inline front/lib/actions/server.ts:92]:
```suggestion
    { ...prodCredentials, groupIds: requestedGroupIds },
```

**Review by flvndvd — APPROVED** [review]

File: connectors/src/connectors/slack/lib/workspace_limits.ts (hunk 6 above)
Comment (lasryaric) [reply connectors/src/connectors/slack/lib/workspace_limits.ts:209]: I talked to the Dust Slack bot using my Slack account and a Slack workflow, both work well 👍

**Pull request merged**